src/edu/washington/escience/myria/coordinator/MasterCatalog.java

```diff
@@ -1494,6 +1504,88 @@ protected Void job(final SQLiteConnection sqliteConnection) throws CatalogExcept
     }
   }
 
+  /*
+   * Mark indexes in the catalog
+   */
+
+  public void markIndexesInCatalog(@Nonnull final RelationKey relation, @Nonnull final List<IndexRef> indexes)
+      throws CatalogException {
+    Objects.requireNonNull(relation, "relation");
+    Objects.requireNonNull(indexes, "indexes");
+    if (isClosed) {
+      throw new CatalogException("Catalog is closed.");
+    }
+
+    /* Do the work */
+    for (int i = 0; i < indexes.size(); i++) {
+      final int indexID = i;
+      try {
+        queue.execute(new SQLiteJob<Void>() {
+          @Override
+          protected Void job(final SQLiteConnection sqliteConnection) throws CatalogException, SQLiteException {
+            try {
+              SQLiteStatement statement =
+                  sqliteConnection
+                      .prepare("UPDATE relation_schema SET is_indexed=1 WHERE user_name=? AND program_name=? AND relation_name=? AND col_index=?;");
+              statement.bind(1, relation.getUserName());
+              statement.bind(2, relation.getProgramName());
+              statement.bind(3, relation.getRelationName());
+              statement.bind(4, indexes.get(indexID).getColumn());
+              statement.stepThrough();
+              statement.dispose();
+              statement = null;
+
+            } catch (final SQLiteException e) {
+              throw new CatalogException(e);
+            }
+            return null;
+          }
+        }).get();
+      } catch (InterruptedException | ExecutionException e) {
+        throw new CatalogException(e);
+      }
+    }
+  }
+
+  /*
+   * Register UDFs in the catalog
+   */
+
+  public void registerUDFs(@Nonnull final String udfName, @Nonnull final String udfDefinition) throws CatalogException {
+    Objects.requireNonNull(udfName, "udf name");
+    Objects.requireNonNull(udfDefinition, "udf definition");
+    if (isClosed) {
+      throw new CatalogException("Catalog is closed.");
+    }
+
+    /* Do the work */
+    try {
+      queue.execute(new SQLiteJob<Void>() {
+        @Override
+        protected Void job(final SQLiteConnection sqliteConnection) throws CatalogException, SQLiteException {
+          try {
+            long udf_id = sqliteConnection.getLastInsertId();
```

Comment:
This seems dicey, even if it's safe given that only one thread is using this connection. I thought our general approach to primary keys (admittedly contradicting general performance-oriented practice) was to avoid unnecessarily synthesizing primary keys (e.g. integer IDs) if we already have a suitable (possibly composite) primary key. Isn't `udf_name` already such a key?